### Import raw data

In [1]:
import pandas as pd
import numpy as np

# 显示所有列
pd.set_option('display.max_columns', None)
df = pd.read_csv("C:/Users/Fan_y/.cache/kagglehub/datasets/vanthanadevi08/water-quality-prediction/versions/1/Water Quality Prediction.csv") #Fan's path
#df = pd.read_csv("../data/raw/Water Quality Prediction.csv")

print(df.head())

   Index        pH      Iron   Nitrate    Chloride           Lead      Zinc  \
0      0  8.332988  0.000083  8.605777  122.799772   3.710000e-52  3.434827   
1      1  6.917863  0.000081  3.734167  227.029851   7.850000e-94  1.245317   
2      2  5.443762  0.020106  3.816994  230.995630   5.290000e-76  0.528280   
3      3  7.955339  0.143988  8.224944  178.129940  4.000000e-176  4.027879   
4      4  8.091909  0.002167  9.925788  186.540872  4.170000e-132  3.807511   

            Color  Turbidity  Fluoride    Copper      Odor     Sulfate  \
0       Colorless   0.022683  0.607283  0.144599  1.626212   87.266538   
1    Faint Yellow   0.019007  0.622874  0.437835  1.686049  144.010981   
2    Light Yellow   0.319956  0.423423  0.431588  3.414619  275.702107   
3  Near Colorless   0.166319  0.208454  0.239451       NaN  147.163233   
4    Light Yellow   0.004867  0.222912  0.616574  0.795310  175.275175   

   Conductivity  Chlorine     Manganese  Total Dissolved Solids  Source  \
0    

In [2]:
import os
os.makedirs("../reports/figures", exist_ok=True)
os.makedirs("../data/processed", exist_ok=True)

### Dataset Info

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 24 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   Index                   1048575 non-null  int64  
 1   pH                      1028344 non-null  float64
 2   Iron                    1041584 non-null  float64
 3   Nitrate                 1029880 non-null  float64
 4   Chloride                1017741 non-null  float64
 5   Lead                    1043891 non-null  float64
 6   Zinc                    1020900 non-null  float64
 7   Color                   1047594 non-null  object 
 8   Turbidity               1039881 non-null  float64
 9   Fluoride                1015357 non-null  float64
 10  Copper                  1013693 non-null  float64
 11  Odor                    1017243 non-null  float64
 12  Sulfate                 1014050 non-null  float64
 13  Conductivity            1019772 non-null  float64
 14  Ch

In [4]:
# Delete Index, because it's useless
df = df.drop(columns=["Index"])
print(f"Shape after dropping Index: {df.shape}")

Shape after dropping Index: (1048575, 23)


In [5]:
df

,pH,Iron,Nitrate,Chloride,Lead,Zinc,Color,Turbidity,Fluoride,Copper,Odor,Sulfate,Conductivity,Chlorine,Manganese,Total Dissolved Solids,Source,Water Temperature,Air Temperature,Month,Day,Time of Day,Target
0,8.332988,8.350000e-05,8.605777,122.799772,3.710000e-52,3.434827,Colorless,0.022683,0.607283,0.144599,1.626212,87.266538,471.683357,3.708178,2.270000e-15,332.118789,NaN,NaN,43.493324,January,29.0,4.0,0
1,6.917863,8.050000e-05,3.734167,227.029851,7.850000e-94,1.245317,Faint Yellow,0.019007,0.622874,0.437835,1.686049,144.010981,432.844908,3.292038,8.020000e-07,284.641984,Lake,15.348981,71.220586,November,26.0,16.0,0
2,5.443762,2.010586e-02,3.816994,230.995630,5.290000e-76,0.528280,Light Yellow,0.319956,0.423423,0.431588,3.414619,275.702107,990.201209,3.560224,7.007989e-02,570.054094,River,11.643467,44.891330,January,31.0,8.0,0
3,7.955339,1.439878e-01,8.224944,178.129940,4.000000e-176,4.027879,Near Colorless,0.166319,0.208454,0.239451,NaN,147.163233,237.028468,3.516907,2.468295e-02,100.043838,Ground,10.092392,60.843233,April,1.0,21.0,0
4,8.091909,2.167128e-03,9.925788,186.540872,4.170000e-132,3.807511,Light Yellow,0.004867,0.222912,0.616574,0.795310,175.275175,385.025855,3.177849,3.296139e-03,168.075545,Spring,15.249416,69.336671,June,29.0,7.0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1048570,8.186428,1.977157e-02,5.968850,115.963429,1.590000e-113,2.277221,Near Colorless,0.020610,1.194165,0.006663,3.149487,43.969084,227.409325,2.718928,2.603963e-02,220.571485,Stream,5.491908,43.817452,March,15.0,14.0,0
1048571,8.046225,1.160000e-05,3.678714,148.053168,5.930000e-25,0.483369,Near Colorless,0.878835,1.203689,0.171610,0.894528,74.133056,124.828663,2.312058,4.427830e-02,406.095969,Lake,10.143768,34.626853,December,16.0,0.0,0
1048572,7.443582,3.103077e-03,7.893399,174.677900,1.080000e-15,1.461659,Near Colorless,0.701053,0.115412,0.641379,3.150154,201.977606,337.630357,3.410297,2.350000e-05,439.086461,Spring,14.612881,55.460415,June,18.0,16.0,0
1048573,6.897232,6.980000e-10,5.757980,125.564223,1.300000e-18,0.804589,Near Colorless,0.156424,0.554729,0.018875,1.997442,204.391754,569.077747,2.803234,3.220000e-19,278.051032,Well,7.351956,65.055864,May,5.0,3.0,0


In [6]:
# Target distribution

target_pct = (
    df["Target"]
    .value_counts(normalize=True)
    * 100
)

print(target_pct)

Target
0    77.041795
1    22.958205
Name: proportion, dtype: float64


### Missing Value

In [7]:
# Overview of missing value
missing = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing %":     (df.isnull().sum() / len(df) * 100).round(2)
})
missing = missing[missing["Missing Count"] > 0].sort_values("Missing %", ascending=False)
print(missing)

                        Missing Count  Missing %
Copper                          34882       3.33
Sulfate                         34525       3.29
Fluoride                        33218       3.17
Odor                            31332       2.99
Chloride                        30834       2.94
Water Temperature               29688       2.83
Conductivity                    28803       2.75
Zinc                            27675       2.64
Time of Day                     20361       1.94
pH                              20231       1.93
Manganese                       19339       1.84
Nitrate                         18695       1.78
Day                             17549       1.67
Month                           16921       1.61
Source                          15535       1.48
Chlorine                        10162       0.97
Turbidity                        8694       0.83
Iron                             6991       0.67
Air Temperature                  5303       0.51
Lead                

#### Analysis of Missing Time Variables

This dataset records water quality measurements, not a time series for forecasting.
Before deciding how to handle missing values in Month / Day / Time of Day,
we first assess whether these variables carry predictive signal.

* **Month** — marginally relevant: summer heat promotes algal growth and rainy seasons
  increase surface runoff. However, this is a synthetic Kaggle dataset, so seasonal
  effects may not be genuine.
* **Day** — nearly meaningless: water quality on the 1st and 15th of a month has no
  systematic difference. This is simply a sampling timestamp.
* **Time of Day** — very weak signal: morning vs. afternoon differences in light/temperature
  are already captured by Water Temperature and Air Temperature, making Time of Day
  largely redundant.


In [8]:
# Check rows where Month is missing
missing_month = df[df["Month"].isna()]
print(f"Number of rows with missing Month: {len(missing_month)}")
print(missing_month.head(10))


Number of rows with missing Month: 16921
           pH          Iron   Nitrate    Chloride           Lead      Zinc  \
95   8.274538  1.967803e-02  3.905648  199.219252   3.170000e-91  0.608931   
185  8.498952  2.130000e-05  6.326846  215.696563   1.330000e-95  2.996623   
321  6.994089  2.617884e-02  3.224344  202.636520   1.730000e-14  4.353077   
348  7.782071  1.180000e-08  8.204868  203.324549  2.870000e-113  0.434184   
371  7.213041  2.108476e-01  4.202989  159.003847  3.940000e-140  1.542676   
423  7.796400  2.362215e-02  5.684404  145.418616   3.670000e-27  1.244731   
543       NaN  7.805703e-03  4.634547  109.515217   5.680000e-05  1.412661   
589  7.682119  8.900000e-07  4.745942  116.771918   1.130000e-10  3.579267   
590  6.535819  4.938256e-03  2.720528  164.570384   2.570000e-44  2.739641   
611  7.025332  8.170000e-12  3.583202  147.512488   1.650000e-93  0.033050   

              Color  Turbidity  Fluoride    Copper      Odor     Sulfate  \
95     Faint Yellow   0.

In [9]:
# Check rows where at least one time column is missing (same batch?)
missing_time = df[df["Month"].isna() | df["Day"].isna() | df["Time of Day"].isna()]
print(f"Rows with at least one missing time field: {len(missing_time)}")

# Missing pattern: do all three columns go missing together?
print("\nMissing combination counts:")
print(missing_time[["Month", "Day", "Time of Day"]].isnull().value_counts())


Rows with at least one missing time field: 53916

Missing combination counts:
Month  Day    Time of Day
False  False  True           19732
       True   False          16927
True   False  False          16347
False  True   True             336
True   False  True             288
       True   False            281
              True               5
Name: count, dtype: int64


In [10]:
# Check whether other columns are also missing in these rows
# (is this a systematic / structural missingness?)
print("Overall missing counts for rows with missing time fields:")
print(missing_time.isnull().sum())


Overall missing counts for rows with missing time fields:
pH                         1062
Iron                        345
Nitrate                     986
Chloride                   1586
Lead                        244
Zinc                       1477
Color                        50
Turbidity                   470
Fluoride                   1740
Copper                     1741
Odor                       1585
Sulfate                    1743
Conductivity               1449
Chlorine                    551
Manganese                  1033
Total Dissolved Solids       15
Source                      798
Water Temperature          1453
Air Temperature             255
Month                     16921
Day                       17549
Time of Day               20361
Target                        0
dtype: int64


In [11]:
# Check Source distribution in rows with missing time fields
# to see whether missingness is concentrated in one source type
print("Source distribution in rows with missing time fields:")
print(missing_time["Source"].value_counts(dropna=False))

# Compare against the full dataset's Source distribution
print("\nSource distribution in the full dataset:")
print(df["Source"].value_counts(dropna=False))


Source distribution in rows with missing time fields:
Source
Ground       6727
Aquifer      6708
Well         6667
Spring       6648
River        6636
Lake         6601
Stream       6598
Reservoir    6533
NaN           798
Name: count, dtype: int64

Source distribution in the full dataset:
Source
Well         129654
Aquifer      129609
Stream       129574
Ground       129433
River        128978
Lake         128637
Reservoir    128582
Spring       128573
NaN           15535
Name: count, dtype: int64


#### Missing Time Variables — Findings

Based on the analysis above, the missingness in the three time fields is **random (MAR)**,
not systematic. The Source distribution in rows with missing time values mirrors the full
dataset, confirming there is no structural bias.

The chemical indicators in these rows show only minor missingness (e.g., pH 1 062 / 53 916 = 2%),
so a missing time field does not indicate that the entire row is low quality.

**Imputation strategy:** random sampling from the observed values of each time column.
This preserves the empirical distribution without introducing the bias that a fixed
median value would create (median imputation would artificially inflate the frequency
of a single value — e.g., making day 16 appear far more often than it truly is).


In [12]:
# Validate correlation between time variables and Target
# Month → numeric (1-12)
month_map = {
    "January": 1, "February": 2, "March": 3,
    "April": 4,   "May": 5,      "June": 6,
    "July": 7,    "August": 8,   "September": 9,
    "October": 10,"November": 11,"December": 12
}
df["Month"] = df["Month"].map(month_map)

# Correlation with Target
time_corr = df[["Month", "Day", "Time of Day", "Target"]].corr()["Target"].drop("Target")
print("Correlation between time variables and Target:")
print(time_corr.round(4))


Correlation between time variables and Target:
Month          0.0001
Day           -0.0003
Time of Day   -0.0014
Name: Target, dtype: float64


#### Time Variable Correlation Results

All three time variables show correlation with Target below 0.05,
confirming they are sampling records rather than meaningful predictors of water quality.
We proceed with random imputation to fill missing values.


#### Missing Values in Categorical Variables

There are only two categorical variables: Color and Source.
Imputing either with a mode or a placeholder category would be logically unsound —
inventing a water source or colour that was not recorded distorts the data.

* Source: 15 535 missing rows — missing rate ≈ 1.48 %
* Color: 981 missing rows — missing rate ≈ 0.09 %

Both rates are low, so we simply drop those rows.


In [13]:
# Fill missing time variables using random sampling from observed values
# (preserves empirical distribution; avoids inflating a single median value)
time_cols = ["Month", "Day", "Time of Day"]

for col in time_cols:
    n_missing = df[col].isna().sum()
    if n_missing > 0:
        observed = df[col].dropna().values
        fill_values = np.random.choice(observed, size=n_missing, replace=True)
        df.loc[df[col].isna(), col] = fill_values
        print(f"{col}: filled {n_missing:,} missing values by random sampling")
    else:
        print(f"{col}: no missing values")

# Drop rows where Color or Source is missing
before = len(df)
df = df.dropna(subset=["Color", "Source"])
after = len(df)
print(f"\nDropped rows with missing Color / Source: {before:,} → {after:,} (lost {before-after:,} rows)")


Month: filled 16,921 missing values by random sampling
Day: filled 17,549 missing values by random sampling
Time of Day: filled 20,361 missing values by random sampling

Dropped rows with missing Color / Source: 1,048,575 → 1,032,071 (lost 16,504 rows)


In [14]:
# Count remaining missing values per column
missing_values = df.isna().sum()
print(missing_values)


pH                        19924
Iron                       6876
Nitrate                   18417
Chloride                  30340
Lead                       4616
Zinc                      27241
Color                         0
Turbidity                  8559
Fluoride                  32690
Copper                    34285
Odor                      30833
Sulfate                   33991
Conductivity              28362
Chlorine                   9985
Manganese                 19026
Total Dissolved Solids      294
Source                        0
Water Temperature         29202
Air Temperature            5220
Month                         0
Day                           0
Time of Day                   0
Target                        0
dtype: int64


#### Missing Values in Numerical Variables

The dataset contains over 1 million rows. Dropping rows with missing numerical values
is cleaner than imputing them — it avoids introducing artificial bias into chemical
measurements that should reflect real sensor readings.


In [15]:
# Drop rows with any remaining missing values (numerical columns)
before = len(df)
df = df.dropna()
after = len(df)
print(f"Dropped rows with missing numerical values: {before:,} → {after:,} (lost {before-after:,} rows)")


Dropped rows with missing numerical values: 1,032,071 → 739,148 (lost 292,923 rows)


#### Remove Unreasonable Temperature Values

Water temperature outside the range **4 °C – 211 °C** is physically implausible
for any water quality monitoring context and most likely reflects sensor errors
or data entry mistakes. Rows outside this range are dropped.
The same filter is applied to Air Temperature for consistency.


In [16]:
# Remove rows with physically implausible temperature values
TEMP_MIN, TEMP_MAX = 4, 211

before = len(df)

df = df[
    (df["Water Temperature"] >= TEMP_MIN) & (df["Water Temperature"] <= TEMP_MAX) &
    (df["Air Temperature"]   >= TEMP_MIN) & (df["Air Temperature"]   <= TEMP_MAX)
]

after = len(df)
print(f"Temperature filter ({TEMP_MIN}–{TEMP_MAX} °C):")
print(f"  Rows before : {before:,}")
print(f"  Rows removed: {before - after:,}")
print(f"  Rows after  : {after:,}")
print(f"\nWater Temperature range after filter: "
      f"{df['Water Temperature'].min():.2f} – {df['Water Temperature'].max():.2f}")
print(f"Air Temperature range after filter  : "
      f"{df['Air Temperature'].min():.2f} – {df['Air Temperature'].max():.2f}")


Temperature filter (4–211 °C):
  Rows before : 739,148
  Rows removed: 4,392
  Rows after  : 734,756

Water Temperature range after filter: 4.00 – 210.82
Air Temperature range after filter  : 4.00 – 144.08


In [17]:
# Step 6: Final validation
assert df.isnull().sum().sum() == 0, "Still has missing values!"

print(f" Cleaning complete")
print(f"Original row count : 1,048,575")
print(f"Cleaned row count  : {len(df):,}")
print(f"Retention rate     : {len(df)/1048575*100:.1f}%")
print(f"\nTarget distribution (after cleaning):")
print(df["Target"].value_counts())
print(df["Target"].value_counts(normalize=True).round(3))
print(f"\nFinal columns ({len(df.columns)}):")
print(df.columns.tolist())



✅ Cleaning complete
Original row count : 1,048,575
Cleaned row count  : 734,756
Retention rate     : 70.1%

Target distribution (after cleaning):
Target
0    565630
1    169126
Name: count, dtype: int64
Target
0    0.77
1    0.23
Name: proportion, dtype: float64

Final columns (23):
['pH', 'Iron', 'Nitrate', 'Chloride', 'Lead', 'Zinc', 'Color', 'Turbidity', 'Fluoride', 'Copper', 'Odor', 'Sulfate', 'Conductivity', 'Chlorine', 'Manganese', 'Total Dissolved Solids', 'Source', 'Water Temperature', 'Air Temperature', 'Month', 'Day', 'Time of Day', 'Target']


#### Cleaning Result

Target distribution: 0.77 vs 0.23, consistent with the raw data.
This confirms that the drop operations did not introduce sampling bias —
the cleaning process was effectively random with respect to the target variable.


In [18]:
# Step 7: Save cleaned dataset
df.to_csv("../data/processed/water_quality_cleaned.csv", index=False)
print("Saved → data/processed/water_quality_cleaned.csv")

Saved → data/processed/water_quality_cleaned.csv
